### PR TITLE
Mod0 week 4 code review 1

### PR DESCRIPTION
**Pull request by RicardoRojo**
Added social networking to the project.

### CODE REVIEW

File: app/controllers/relationships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  def destroy
+    relationship = Relationship.find(params[:id])
+    relationship.destroy if relationship.follower == current_user
+    redirect_to people_path
+  end
+
+  def create
+    leader = User.find(params[:leader_id])
+    Relationship.create(leader: leader,
+     follower: current_user) unless current_user.not_followable?(leader)
```
Comment (tomtomecek) [inline app/controllers/relationships_controller.rb:17]: This should be styled differently, keys below each other aligned or both aligned and indented below upper line like

``` ruby
Relationship.create(
  leader: leader,
  follower: current_user
) unless current_user.not_followable?(leader)
```

File: app/models/relationship.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Relationship < ActiveRecord::Base
+  validates :follower_id, presence: true
+  validates :leader_id, presence: true
```
Comment (tomtomecek) [inline app/models/relationship.rb:3]: This certainly should not be here. Why are you validating Relationship? Is user inputting data here directly? You put validations only there where user has direct impact.

File: app/models/user.rb [hunk 3]
```diff
@@ -15,4 +18,16 @@ def normalize_queue_item_list
   def queued_video?(video)
     queue_items.find_by(video: video)
   end
+
+  def already_followed?(leader)
+    following_relationships.map(&:leader).include?(leader)
+  end
+
+  def is_the?(role)
```
Comment (tomtomecek) [inline app/models/user.rb:26]: `is_the?` is a little bit strange method name, i would probably do something else like `has_role?`.

File: app/models/user.rb [hunk 4]
```diff
@@ -15,4 +18,16 @@ def normalize_queue_item_list
...
+  def already_followed?(leader)
+    following_relationships.map(&:leader).include?(leader)
+  end
+
+  def is_the?(role)
+    self == role
+  end
+
+  def not_followable?(user)
+    self.already_followed?(user) || self.is_the?(user)
```
Comment (tomtomecek) [inline app/models/user.rb:31]: Its hard to follow this - `self.is_the?(user)` first, you pass in user, but when I look at the `is_the?` method user changes to role.

Comment (tomtomecek) [inline app/models/user.rb:31]: You don't need to use `self` in any of these cases.

File: app/views/relationships/index.html.haml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+      %tr
+        %th(width="30%")
+        %th(width="20%") Videos in Queue
+        %th(width="20%") Followers
+        %th(width="30%") Unfollow
+    %tbody
+      - @relationships.each do |relationship|
+        %tr
+          %td
+            %img(src="http://www.gravatar.com/avatar/#{Digest::MD5.hexdigest('#{relationship.leader.email.downcase}')}?s=40")
```
Comment (tomtomecek) [inline app/views/relationships/index.html.haml:15]: This gravatar URL should definitely be a helper method. Too hard to digest/read at this current state. Also I would recommend using `= image_tag`

File: app/views/shared/_header.html.haml [hunk 6]
```diff
@@ -5,7 +5,7 @@
     %ul.col-md-4.clearfix
       %li= link_to "Videos"
       %li= link_to "My Queue"
-      %li= link_to "People"
+      %li= link_to "People", people_path
     .col-md-5.navbar-form
       = form_tag( "/videos/search", method: "get") do
```
Comment (tomtomecek) [inline app/views/shared/_header.html.haml:10]: I am not understand the space after `(`.

File: app/views/users/show.html.haml [hunk 7]
```diff
@@ -0,0 +1,36 @@
+%section.user.container
+  .row
+    .col-sm-10.col-sm-offset-1
+      %article
+        %header
+          %img(src="http://www.gravatar.com/avatar/#{Digest::MD5.hexdigest(@user.email.downcase)}?s=40")
```
Comment (tomtomecek) [inline app/views/users/show.html.haml:6]: Same here for gravatar URL helper.

File: app/views/users/show.html.haml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    .col-sm-10.col-sm-offset-1
+      %header
+        %h3 #{@user.full_name}'s Reviews (#{@user.reviews.count})
+      %ul
+        - @user.reviews.each do |review|
+          %article.review
+            %li.row
+              .col-sm-2
+                %p
+                  = link_to "\"#{review.video.title}\""
```
Comment (tomtomecek) [inline app/views/users/show.html.haml:33]: You don't need to use quotes here `= link_to review.video.title` is enough.

File: app/views/videos/show.html.haml [hunk 9]
```diff
@@ -39,6 +39,6 @@
                 %span Rating: #{review.rating} / 5
                 %p 
                   by
-                  = link_to "#{review.user.full_name}"
+                  = link_to "#{review.user.full_name}", user_path(review.user)
```
Comment (tomtomecek) [inline app/views/videos/show.html.haml:42]: `link_to review.user.full_name, review.user`

File: spec/controllers/relationships_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+  describe "POST create" do
+    it_behaves_like "require user" do
+      let(:action) {post :create, leader_id: 1}
+    end
+
+    let(:alice) {Fabricate(:user)}
+    let(:bob) {Fabricate(:user)}
+
+    before { set_user(alice)}
```
Comment (tomtomecek) [inline spec/controllers/relationships_controller_spec.rb:59]: There should be a space before `}`.

File: spec/controllers/relationships_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+      let(:action) {delete :destroy, id: 1}
+    end
+
+    let(:alice) {Fabricate(:user)}
+    let(:bob) {Fabricate(:user)}
+    let(:charlie) {Fabricate(:user)}
+
+    before do
+      set_user(alice)
+    end
```
Comment (tomtomecek) [inline spec/controllers/relationships_controller_spec.rb:30]: This could rather be 1 liner. `before { set_user(alice) }`

File: spec/controllers/users_controller_spec.rb [hunk 12]
```diff
@@ -64,4 +64,19 @@
       end
     end
   end
+
+  describe "Get show" do
```
Comment (tomtomecek) [inline spec/controllers/users_controller_spec.rb:68]: Everywhere else you caps lock the HTTP Verb but not here. why?

File: spec/features/social_networking.rb [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+# login
+# home
+# click on video
+# click on reviews
+# social page of user
+# follow
+# people includes user
+# unfollow
+# person not in the list
```
Comment (tomtomecek) [inline spec/features/social_networking.rb:11]: This is OK as notes when you build the feature, but not for code review.

File: app/controllers/relationships_controller.rb (hunk 1 above)
Comment (RicardoRojo) [inline app/controllers/relationships_controller.rb:17]: True, this way is clearer. Thanks Tom

File: app/models/user.rb (hunk 3 above)
Comment (RicardoRojo) [inline app/models/user.rb:26]: :). Yep, i know, this name really sucks. Sometimes just cant find a name that i like.

File: app/models/user.rb (hunk 4 above)
Comment (RicardoRojo) [inline app/models/user.rb:31]: true

File: app/views/relationships/index.html.haml (hunk 5 above)
Comment (RicardoRojo) [inline app/views/relationships/index.html.haml:15]: Modified. I created a helper method.

File: app/views/shared/_header.html.haml (hunk 6 above)
Comment (RicardoRojo) [inline app/views/shared/_header.html.haml:10]: Just a typo error.

File: app/views/videos/show.html.haml (hunk 9 above)
Comment (RicardoRojo) [inline app/views/videos/show.html.haml:42]: True. I have done this a thousand times but keep making the same error. Sorry. I think i have to concentrate more and stop thinking in the next thing i have to do before ending the current one.

File: spec/controllers/relationships_controller_spec.rb (hunk 11 above)
Comment (RicardoRojo) [inline spec/controllers/relationships_controller_spec.rb:30]: true

File: spec/controllers/users_controller_spec.rb (hunk 12 above)
Comment (RicardoRojo) [inline spec/controllers/users_controller_spec.rb:68]: Just lack of concentration again. Sorry.

File: spec/features/social_networking.rb (hunk 13 above)
Comment (RicardoRojo) [inline spec/features/social_networking.rb:11]: doh. Forgot to delete.
